nkms/characters.py

```diff
@@ -372,33 +400,64 @@ def __init__(self, *args, **kwargs):
         from nkms.policy.models import WorkOrderHistory  # Need a bigger strategy to avoid circulars.
         self._saved_work_orders = WorkOrderHistory()
 
-    def follow_treasure_map(self, hrac):
-        for ursula_interface_id in self.treasure_maps[hrac]:
-            if ursula_interface_id in self.known_nodes:
-                # If we already know about this Ursula,
-                # we needn't learn about it again.
-                continue
-
-            # TODO: perform this part concurrently.
-            value = self.server.get_now(ursula_interface_id)
-
-            # TODO: Make this much prettier
-            header, signature, ursula_pubkey_sig, _hrac, (
-                port, interface, ttl) = dht_value_splitter(value, msgpack_remainder=True)
-
-            if header != constants.BYTESTRING_IS_URSULA_IFACE_INFO:
-                raise TypeError("Unknown DHT value.  How did this get on the network?")
-
-            # TODO: If we're going to implement TTL, it will be here.
-            self.known_nodes[ursula_interface_id] = \
-                Ursula.as_discovered_on_network(
-                    dht_port=port,
-                    dht_interface=interface,
-                    powers_and_keys=({SigningPower: ursula_pubkey_sig})
-                )
-
-    def get_treasure_map(self, policy, networky_stuff, using_dht=False):
-        map_id = policy.treasure_map_dht_key()
+    def follow_treasure_map(self, hrac, using_dht=False):
+
+        treasure_map = self.treasure_maps[hrac]
+        number_of_known_treasure_ursulas = 0
+        if not using_dht:
+            for ursula_interface_id in treasure_map:
+                pubkey = UmbralPublicKey.from_bytes(ursula_interface_id)
+                if pubkey in self.known_nodes:
+                    number_of_known_treasure_ursulas += 1
+
+            newly_discovered_nodes = {}
+            nodes_to_check = iter(self.known_nodes.values())
+
+            while number_of_known_treasure_ursulas < treasure_map.m:
+                try:
+                    node_to_check = next(nodes_to_check)
```

Comment:
Lots of Python magic happening in this function. Can we document the exact algorithm how we traverse this in plain English somewhere? Perhaps in the wiki?

Reply:
Hmm - so this isn't traversal; this is Bob discovering new nodes until he has >= `m`.  I don't think there's anything resembling "magic" here, although I agree that it can use at least a docstring.